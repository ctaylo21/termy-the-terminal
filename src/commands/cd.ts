import get from 'lodash/get';
import has from 'lodash/has';
import {
  AutoCompleteResponse,
  CommandResponse,
  FileSystem,
  ItemListType,
} from '../index';
import {
  convertInternalPathToExternal,
  getInternalPath,
} from './utilities/index';
import autoComplete from './autoComplete';

/**
 * Given a file system, validates if changing directories from a given path
 * to a new path is possible, and returns the new path if so.
 *
 * @param fileSystem {object} - filesystem to cd upon
 * @param currentPath {string} - current path within filesystem
 * @param targetPath  {string} - path to change to
 * @returns Promise<object> - resolves with new path if successful, rejects if not
 */
function cd(
  fileSystem: FileSystem,
  currentPath: string,
  targetPath: string,
): Promise<CommandResponse> {
  return new Promise((resolve, reject): void => {
    if (!targetPath) {
      reject('path can not be empty.');
    }
    const internalCdPath = getInternalPath(currentPath, targetPath);

    if (!internalCdPath) {
      resolve({
        updatedState: {
          currentPath: '/',
        },
      });
    }

    if (
      has(fileSystem, internalCdPath) &&
      get(fileSystem, internalCdPath).type !== 'FILE'
    ) {
      resolve({
        updatedState: {
          currentPath: convertInternalPathToExternal(internalCdPath),
        },
      });
    }

    reject(`path does not exist: ${targetPath}`);
  });
}

/**
 * Given a fileysystem, current path, and target, list the items in the desired
 * folder that start with target string
 *
 * @param fileSystem {object} - filesystem to ls upon
 * @param currentPath {string} - current path within filesystem
 * @param target {string} - string to match against (maybe be path)
 * @returns Promise<object> - resolves with contents that match target in path
 */
function cdAutoComplete(
  fileSystem: FileSystem,
  currentPath: string,
  target: string,
): Promise<AutoCompleteResponse> {
  const filterNonFilesFn = (item: ItemListType): boolean =>
    item[Object.keys(item)[0]].type === 'FOLDER';

  return autoComplete(fileSystem, currentPath, target, filterNonFilesFn);
}

export default {
  handler: cd,
  autoCompleteHandler: cdAutoComplete,
};
