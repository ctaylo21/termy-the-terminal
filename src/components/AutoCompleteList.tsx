import React from 'react';

export interface AutoCompleteList {
  [index: string]: {
    type: 'FOLDER' | 'FILE';
  };
}

interface AutoCompleteListProps {
  items: AutoCompleteList;
  activeItemIndex?: number;
}

const AutoCompleteList: React.FC<AutoCompleteListProps> = (
  props,
): JSX.Element => {
  const { items, activeItemIndex } = props;

  const autoCompleteItems = Object.keys(items).map(
    (key, index): JSX.Element => {
      if (items[key].type === 'FOLDER') {
        return (
          <span
            className={
              activeItemIndex === index
                ? 'ls-preview-folder active'
                : 'ls-preview-folder'
            }
            key={key}
          >
            {key}/
          </span>
        );
      }

      return (
        <span
          className={
            activeItemIndex === index
              ? 'ls-preview-file active'
              : 'ls-preview-file'
          }
          key={key}
        >
          {key}
        </span>
      );
    },
  );

  return <div className="preview-list">{autoCompleteItems}</div>;
};

export default AutoCompleteList;
