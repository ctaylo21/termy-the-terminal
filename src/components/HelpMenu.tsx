import React from 'react';
import TerminalContext from '../context/TerminalContext';

export const HelpMenu: React.FC<{}> = (): JSX.Element => {
  return (
    <TerminalContext.Consumer>
      {(commands): JSX.Element => (
        <div id="help-container">
          <ul aria-label="help-menu">
            {Object.keys(commands)
              .filter((command) => commands[command].description)
              .map(
                (command): JSX.Element => (
                  <li key={command}>
                    <span className="help-command-name">{command}</span> -{' '}
                    <span>{commands[command].description}</span>
                  </li>
                ),
              )}
          </ul>
        </div>
      )}
    </TerminalContext.Consumer>
  );
};

export default HelpMenu;
