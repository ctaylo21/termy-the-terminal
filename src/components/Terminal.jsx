import React, { Component } from 'react';
import PropTypes from 'prop-types';
import History from './History';
import Input from './Input';

export default class Terminal extends Component {
  constructor(props) {
    super(props);
    this.currentCommandId = 0;
    this.state = {
      history: [],
      inputValue: '',
    };
  }

  handleKeyPress = (event) => {
    if (event.key === 'Enter') {
      const { history } = this.state;
      const updatedHistory = history.concat({
        id: this.currentCommandId++,
        value: event.target.value,
      });

      this.setState({
        history: updatedHistory,
        inputValue: '',
      });
    }
  }

  render() {
    const { history, inputValue } = this.state;
    const { inputPrompt } = this.props;
    return (
      <>
        <History history={history} />
        <Input inputPrompt={inputPrompt} handleKeyPress={this.handleKeyPress} inputValue={inputValue} />
      </>
    );
  }
}

Terminal.propTypes = {
  inputPrompt: PropTypes.string.isRequired,
};
