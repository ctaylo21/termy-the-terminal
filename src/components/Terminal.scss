@import '~reset-css';
@import '../styles/vars';

#terminal-wrapper {
  background-color: map-get($themeDark, background);
  box-sizing: content-box;
  color: map-get($themeDark, fontColor);
  font-family: $fontFamily;
  font-size: $fontSize;
  height: 100%;
  padding: 0 1em;

  #history-container {
    li {
      margin-bottom: 1em;
    }
  }

  #input-container {
    padding-top: 1em;

    form {
      display: flex;

      input {
        background: none;
        border: none;
        color: map-get($themeDark, fontColor);
        font-family: $fontFamily;
        font-size: $fontSize;
        outline: none;
        -webkit-appearance: none;

        flex-grow: 100;
        margin: 0 10px;
      }

      #inputPromptChar {
        color: map-get($themeDark, green);
      }
    }
  }
}
