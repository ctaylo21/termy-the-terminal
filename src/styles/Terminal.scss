@import './vars';

#terminal-wrapper {
  background-color: map-get($themeDark, background);
  box-sizing: content-box;
  color: map-get($themeDark, fontColor);
  font-family: map-get($themeDark, fontFamily);
  font-size: map-get($themeDark, fontSize);
  height: 100%;
  overflow: auto;
  padding: 0 1em;

  #input-container {
    padding: 1em 0;

    form {
      display: flex;

      input {
        background: none;
        border: none;
        color: map-get($themeDark, fontColor);
        font-family: map-get($themeDark, fontFamily);
        font-size: map-get($themeDark, fontSize);
        outline: none;
        -webkit-appearance: none;

        flex-grow: 100;
        margin: 0 10px;
      }

      #inputPrompt {
        color: map-get($themeDark, green);
      }
    }
  }

  #history-container {
    & > ul > li > #input-container {
      padding-bottom: 0;
    }

    img {
      margin-top: 10px;
    }

    #help-container ul > li {
      margin: 5px 0;

      & > span.help-command-name {
        color: map-get($themeDark, purple);
      }
    }

    ul {
      list-style-type: none;
      margin: 0;
      padding: 0;
    }

    ul.terminal-ls-list {
      margin: 10px 0 0 10px;

      li {
        display: flex;
        align-items: center;

        svg {
          margin-right: 20px;
        }

        &.ls-folder {
          color: map-get($themeDark, orange);

          svg {
            fill: map-get($themeDark, orange);
            height: 24px;
            width: 24px;
          }
        }

        &.ls-file {
          color: map-get($themeDark, blue);

          svg {
            fill: map-get($themeDark, blue);
            margin-top: 2px;
            margin-bottom: 2px;
            height: 20px;
            width: 24px;
          }
        }
      }
    }
  }

  .preview-list {
    span {
      margin: 0px 10px;

      &.auto-preview-folder {
        color: map-get($themeDark, orange);
      }

      &.auto-preview-file {
        color: map-get($themeDark, blue);
      }

      &.active {
        background-color: map-get($themeDark, highlight);
      }
    }


  }
}
