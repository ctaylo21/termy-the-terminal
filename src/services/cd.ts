import get from 'lodash/get';
import has from 'lodash/has';
import { FileSystem } from '../components/Terminal';

/**
 * Takes a valid Unix path and converts it into a format
 * that matches the internal data structure. Including replacing "/"
 * with ".children" and removing leading slash
 *
 * usr/home/test becomes usr.children.home.children.test
 *
 * @param {string} pathStr - path string to convert
 * @returns {string} - converted string
 */
function convertPathToInternalFormat(pathStr: string): string {
  return pathStr
    .replace(/^\/+/g, '')
    .split('/')
    .map((elem, index, arr) => {
      if (elem !== '..' && index !== arr.length - 1) {
        elem += '.children';
      }
      return elem;
    })
    .join('.');
}

/**
 * Takes a unix path that may contain ".." and return a new
 * string that respects the ".." path segments i.e.
 * /home/user/.. => /home
 *
 * @param pathStr {string} - unix-formatted path
 * @returns {string} - new path after taking ".." into account
 */
function handleDotDotInPath(pathStr: string): string {
  let currentDotDots = 0;
  let pathArr = pathStr.split('/').filter(path => path.length > 0);
  for (let i = pathArr.length - 1; i >= 0; i--) {
    if (pathArr[i] === '..') {
      currentDotDots++;
    } else {
      if (currentDotDots > 0) {
        pathArr.splice(i, 1);
        currentDotDots--;
      }
    }
  }

  return pathArr.filter(path => path !== '..').join('/');
}

/**
 * Takes an path formatted for internal use and converts it to an
 * external format. i.e.
 * home.children.user => /home/user
 *
 * @param pathStr {string} - internally formatted path
 * @returns {string} - path string formatted for terminal use
 */
function convertInternalPathToExternal(pathStr: string): string {
  return (
    '/' +
    pathStr
      .split('.')
      .filter(path => path !== 'children')
      .join('/')
  );
}

/**
 * Given a fileysystem, validates if changing directories from a given path
 * to a new path is possible, and returns the new path if so.
 *
 * @param fileSystem {object} - filesystem to cd upon
 * @param currentPath {string} - current path within filesystem
 * @param pathToCd  {string} - path to change to
 * @returns Promise<string> - resolves with new path if successful, rejects if not
 */
export default function cd(
  fileSystem: FileSystem,
  currentPath: string,
  pathToCd: string,
): Promise<string> {
  return new Promise((resolve, reject) => {
    if (!pathToCd) {
      reject('Path can not be empty.');
    }

    // If current path is anything other than the root, add trailing slash
    const normalizedCurrentPath =
      currentPath === '/' ? currentPath : `${currentPath}/`;

    let internalCdPath = convertPathToInternalFormat(
      handleDotDotInPath(normalizedCurrentPath + pathToCd),
    );

    if (!internalCdPath) {
      resolve('/');
    }

    if (
      has(fileSystem, internalCdPath) &&
      get(fileSystem, internalCdPath).type !== 'FILE'
    ) {
      resolve(convertInternalPathToExternal(internalCdPath));
    }

    reject('Path does not exist.');
  });
}
