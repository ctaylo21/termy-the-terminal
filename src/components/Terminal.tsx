import React, { ChangeEvent, Component, FormEvent } from 'react';
import { History } from './History';
import Input from './Input';
import { cd, ls } from '../services';
import './Terminal.scss';

export class Terminal extends Component<TerminalProps, TerminalState> {
  public readonly state: TerminalState = {
    currentCommandId: 0,
    currentPath: '/',
    history: [],
    inputValue: '',
    promptChar: 'â†’',
  };

  private handleChange = (event: ChangeEvent<HTMLInputElement>): void => {
    this.setState({
      inputValue: event.target.value,
    });
  };

  private handleSubmit = async (
    event: FormEvent<HTMLFormElement>,
  ): Promise<void> => {
    event.preventDefault();

    const commandArgs = this.state.inputValue.split(' ');

    const { history, inputValue } = this.state;

    let result = '';
    switch (commandArgs[0]) {
      case 'cd':
        try {
          const newPath = await cd(
            this.props.fileSystem,
            this.state.currentPath,
            commandArgs[1],
          );

          result = 'cd success';
          this.setState({
            currentPath: newPath,
          });
        } catch (cdException) {
          result = cdException;
        }
        break;
      case 'pwd':
        result = this.state.currentPath;
        break;
      case 'ls':
        const pathToLs = commandArgs[1]
          ? commandArgs[1].startsWith('/')
            ? commandArgs[1]
            : `${this.state.currentPath}/${commandArgs[1]}`
          : this.state.currentPath;

        try {
          const test = await ls(this.props.fileSystem, pathToLs);
          result = JSON.stringify(test);
        } catch (e) {
          result = e;
        }
        break;
      default:
        result = 'Invalid command';
        break;
    }

    const updatedHistory = history.concat({
      id: this.state.currentCommandId,
      result,
      value: inputValue,
    });

    this.setState({
      currentCommandId: this.state.currentCommandId + 1,
      history: updatedHistory,
      inputValue: '',
    });
  };

  public render(): JSX.Element {
    const { currentPath, history, inputValue, promptChar } = this.state;
    return (
      <>
        <History history={history} />
        <Input
          currentPath={currentPath}
          handleChange={this.handleChange}
          handleSubmit={this.handleSubmit}
          inputValue={inputValue}
          promptChar={promptChar}
        />
      </>
    );
  }
}
