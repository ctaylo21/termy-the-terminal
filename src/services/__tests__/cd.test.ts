import cd from '../cd';
import { FileSystem } from '../../components/Terminal';

const testSystem: FileSystem = {
  home: {
    children: {
      folder1: {
        type: 'FOLDER',
        children: {
          folder2: {
            type: 'FOLDER',
            children: {
              file1: {
                type: 'FILE',
                children: null,
              },
            },
          },
        },
      },
    },
    type: 'FOLDER',
  },
};

describe('cd suite', (): void => {
  describe('valid cases', (): void => {
    describe('from root', (): void => {
      test('1 level path', async (): Promise<string> => {
        const fileSystem: FileSystem = {
          home: {
            children: null,
            type: 'FOLDER',
          },
        };

        return expect(cd(fileSystem, '/', 'home')).resolves.toEqual('/home');
      });

      test('multi-level cd', async (): Promise<string> => {
        return expect(
          cd(testSystem, '/', 'home/folder1/folder2'),
        ).resolves.toEqual('/home/folder1/folder2');
      });

      test('.. above root level', async (): Promise<string> => {
        return expect(cd(testSystem, '/', '..')).resolves.toEqual('/');
      });
    });

    describe('from nested path', (): void => {
      test('1 level cd', async (): Promise<string> => {
        return expect(cd(testSystem, '/home', 'folder1')).resolves.toEqual(
          '/home/folder1',
        );
      });

      test('.. 1 level to root', async (): Promise<string> => {
        return expect(cd(testSystem, '/home', '..')).resolves.toEqual('/');
      });

      test('.. 1 level', async (): Promise<string> => {
        return expect(
          cd(testSystem, '/home/folder1/folder2', '..'),
        ).resolves.toEqual('/home/folder1');
      });

      test('.. multiple levels', async (): Promise<string> => {
        return expect(
          cd(testSystem, '/home/folder1/folder2', '../..'),
        ).resolves.toEqual('/home');
      });

      test('.. multiple levels in separate paths', (): Promise<
        Promise<string>
      > => {
        return expect(
          cd(testSystem, '/home/folder1/folder2', '../folder2/../../'),
        ).resolves.toEqual('/home');
      });
    });
  });

  describe('invalid cases', (): void => {
    test('empty path', async (): Promise<string> => {
      const fileSystem: FileSystem = {
        home: {
          children: null,
          type: 'FOLDER',
        },
      };
      return expect(cd(fileSystem, 'path', '')).rejects.toMatchSnapshot();
    });

    test('nested cd to a file', async (): Promise<string> => {
      return expect(
        cd(testSystem, 'path', 'home/folder1/folder2/file1'),
      ).rejects.toMatchSnapshot();
    });
  });
});
