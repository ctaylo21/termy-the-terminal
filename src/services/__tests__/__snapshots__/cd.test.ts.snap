// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cd suite invalid cases empty path 1`] = `"path can not be empty."`;

exports[`cd suite invalid cases nested cd to a file 1`] = `"path does not exist."`;
